hoomd/wall.py

```diff
@@ -0,0 +1,396 @@
+
+    Note:
+        `Cylinder` objects are immutable.
+
+    Args:
+        radius (`float`, optional):
+            The radius of the circle faces of the cylinder.
+        origin (`tuple` [`float`,`float`,`float`], optional):
+            The origin of the cylinder defined as the center of the bisecting
+            circle along the cylinder's axis.
+        axis (`tuple` [`float`,`float`,`float`], optional):
+            A vector perpendicular to the circular faces.
+        inside (`bool`, optional):
+            Whether particles are restricted to the space inside or outside the
+            cylinder.
+
+    Attributes:
+        radius (float):
+            The radius of the circle faces of the cylinder.
+        origin (`tuple` [`float`,`float`,`float`]):
+            The origin of the cylinder defined as the center of the bisecting
+            circle along the cylinder's axis.
+        axis (`tuple` [`float`,`float`,`float`]):
+            A vector perpendicular to the circular faces.
+        inside (bool):
+            Whether particles are restricted to the space inside or outside the
+            cylinder.
+    """
+
+    def __init__(self,
+                 radius=0.0,
+                 origin=(0.0, 0.0, 0.0),
+                 axis=(0.0, 0.0, 1.0),
+                 inside=True):
+        param_dict = ParameterDict(radius=float,
+                                   origin=(float, float, float),
+                                   axis=(float, float, float),
+                                   inside=bool)
+        param_dict["radius"] = radius
+        param_dict["origin"] = origin
+        param_dict["axis"] = axis
+        param_dict["inside"] = inside
+        self._param_dict = param_dict
+
+    def __str__(self):
+        """A string representation of the Cylinder."""
+        return self.__repr__()
+
+    def __repr__(self):
+        """A string representation of the Cylinder."""
+        return f"Cylinder(radius={self.radius}, origin={self.origin}, "
+        f"axis={self.axis}, inside={self.inside})"
+
+    def to_dict(self):
+        """Return a dictionary specifying the cylinder."""
+        return {
+            "radius": self.radius,
+            "origin": self.origin,
+            "axis": self.axis,
+            "inside": self.inside
+        }
+
+
+class Plane(WallGeometry):
+    """Define a Plane in 3D Euclidean space.
+
+    Note:
+        `Plane` objects are immutable.
+
+    Args:
+        origin (`tuple` [`float`,`float`,`float`], optional):
+            A point that lies on the plane used with ``normal`` to fully specify
+            the plane.
+        normal (`tuple` [`float`,`float`,`float`], optional):
+            The normal vector to the plane.
+        inside (`bool`, optional):
+            Whether particles are restricted to the space inside or outside the
+            cylinder. Inside is the side of the plane the normal points to, and
+            outside constitutes the other side.
```

Comment:
Plane walls should not have an `inside` parameter. Note that the "inside" of a plane is the side that the normal points towards.

Perhaps we should consider inverting this logic in the code? Computer graphics literature and software typically define outward facing normals.

Reply:
That is true, we just need to be clear in our documentation which half-space the normal points to. Here I don't know if it makes much difference since it is about which direction the wall is active in which is not a typical use case.

Reply:
We should remove barriers to interoperation so that building walls in 3D graphics tools (e.g. blender) can be a typical use case.

Reply:
My only counter argument to inverting the logic would be that in computer graphics it makes sense for the normals to be outward facing because you're typically thinking of the polygons that describe a discrete object that interacts with the world/light simulations outwardly. I feel like it's really a matter of perspective here that we're usually thinking of a confinement situation where we'd be inside of that object. 

However obviously it's just a case of documenting the choice we make. If @joaander is envisioning interoperability with general 3D renders as a common use with walls then I'd tend to agree going with their inside definition makes more sense.

Reply:
@b-butler I actually make the argument here https://github.com/glotzerlab/hoomd-blue/pull/1020#discussion_r748693794 that it is a common use case for some potentials. So I would argue it is important to keep and clearly define.

Reply:
I should clarify, I meant there isn't a standard direction for the normal in the case of confining walls in software that I know of since this is much more niche than graphics not that the ability to use each half-space is not useful.